internal_ws/ykcompile/src/lib.rs

```diff
@@ -1358,8 +1366,12 @@ impl TraceCompiler {
         // Reserved stack space for spilled locals during execution.
         let soff = self.stack_builder.size();
         // Reserved memory on the stack to spill live locals to during a guard failure.
-        let live_off = REG_POOL.len() * 8;
+        let live_off = REG_POOL.len() * REG_SIZE;
 
+        // FIXME Can we re-review this whole comment, as I'm sure it could be made shorter, and I'm
+        // not sure it's right to say that the stack must be 16-byte aligned *before* the call. I
+        // thought it was right *after* the call (once the RV has been pushed).
+        //
```

Comment:
To expand on this, here's a snippet from my notes:

```
> The end of the input argument area shall be aligned on a 16 byte boundary.
In other words, the value(%rsp+ 8)is always a multiple of 16 when control is 
transferred to the function entry point. The stack pointer,%rsp, always
points to the end of the latest allocated stack frame.

https://uclibc.org/docs/psABI-x86_64.pdf
```

So this means that once the return value is pushed (once the call instruction has finished, and the rsp points at the first instr of the callee) the stack is 16-byte aligned, but not *before* the call.

The wording is confusing, but the critical detail is (I think) "end of the input argument area" -- i.e. right before the RV is pushed.

Reply:
I thought we have put this topic to rest. But the fact we are still discussing this is a testament to how awful that part of the SYS-V documentation is written. Did you notice that your two paragraphs contradict each other? Your first one says we need to align to 16 bytes after the return value, and your second paragraph says it needs to be before.

How I understood the documentation (and we confirmed this by looking it the assembler of some test programs) is that we need to align to 16 bytes immediately **before** we do the call. This means that the we are aligned to 16 bytes at the end of the input argument area. The call then pushes the return value onto the stack, misaligning it by 8 bytes. However that means that `%rsp + 8` is a multiple of 16 bytes, like the rest of the documentation specifies.

They should definitely fix those docs, e.g. "In other words, the value(%rsp+ 8)is always a multiple of 16 ~~when~~ before control is transferred to the function entry point."

Reply:
Is it possible to propose changes to the sys-v docs or is it hidden behind an impenetrable standards organisation?

Reply:
Seems it's on Gitlab: https://gitlab.com/x86-psABIs/x86-64-ABI/-/blob/master/x86-64-ABI/low-level-sys-info.tex#L481

Though, before raising a PR they ask you to discuss the changes on their mailing list. I can give that a go.

Reply:
I think it's worth putting a small amount of effort into this, if nothing else to save us heartache in the future :) But if there's much resistance to a change, I'd probably give up: some standards committees are welcoming of changes, some aren't.

Reply:
> I think it's worth putting a small amount of effort into this.

Agreed.